pkg/platform/runtime/buildplan/buildplan.go

```diff
@@ -268,15 +302,18 @@ func NewNamedMapFromBuildPlan(build *model.Build) (artifact.NamedMap, error) {
 	return res, nil
 }
 
-// BuildtimeArtifacts iterates through all artifacts in a given build and
+// NewBuildtimeMap iterates through all artifacts in a given build and
 // adds the artifact's dependencies to a map. This is different from the
 // runtime dependency calculation as it includes ALL of the input artifacts of the
 // step that generated each artifact. The includeBuilders argument determines whether
 // or not to include builder artifacts in the final result.
-func BuildtimeArtifacts(build *model.Build, includeBuilders bool) (artifact.Map, error) {
-	result := make(artifact.Map)
-	lookup := make(map[strfmt.UUID]interface{})
+func NewBuildtimeMap(build *model.Build) (artifact.Map, error) {
```

Comment:
This is my second run reading through this code and I'll apologize in advance for not raising this the first time around;

It feels a little unproductive for us to iterate and convert the build map twice through two distinct code paths to calculate the artifact map. Once for runtime artifacts, and then once for build time artifacts. In each case the logic is very similar.

Would it be better if we did a single calculation of the artifact map, and the artifact map itself simply has a property like `IsBuildTimeOnly` which can be further filtered on?

That'd keep the performance nimble but also reduces the amount of code to maintain and update.

Reply:
I was able to use one function with a boolean switch. Unfortunately, we can't filter the build plan down from the runtime closure to the buildtime closure because the runtime closure is a subset of the buildtime one. This would mean that we would have to _always_ calculate the buildtime closure and then filter down to the runtime one.

Instead, I've opted to just use one function to calculate both and use a boolean switch. This does mean that there are instances where we may have to iterate through the build plan multiple times to calculate both closures.

Reply:
I don't fully understand this limitation. We wouldn't be filtering down from either runtime or buildtime closure. We would be iterating through the buildplan and generating both based (if requested). The point being there is only a single iteration that produces both a build time and runtime closure.